### PR TITLE
Fix SDPA sliding window compatibility

### LINKED ISSUE #28980: Add sliding window attention to sdpa in mistral
### Feature request

https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py#L1006-L1023
image

In the code listed above, the latest version of transformers cannot use sliding window feature in mistral model.
I doubt that the reason is you mentioned above,
https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py#L687-L688
image
And this issue in PyTorch makes you bugged with custom attn_mask like sliding window attention mask.
https://github.com/pytorch/pytorch/issues/112577

While this issue has been fixed since torch 2.2.0, and it has been released two weeks ago, can you add this feature back to sdpa kernel in mistral?

### Motivation

I cannot use sliding window with sdpa right now, cause my gpu card is V100, i cannot work with flashattention2.

### Your contribution

I think we can pass sliding_window param to _prepare_4d_causal_attention_mask_for_sdpa function.

### PR DESCRIPTION
**Pull request by fxmarty**
As per title, fixes https://github.com/huggingface/transformers/issues/28980

Supersedes https://github.com/huggingface/transformers/pull/29220 https://github.com/huggingface/transformers/pull/29407 as the implementation ends up being different (added you as co-author here @ehuaa).

This bug dates back to https://github.com/huggingface/transformers/pull/26572 where `sliding_window` was not properly accounted for in the `_prepare_4d_causal_attention_mask_for_sdpa` method. Since then, SDPA support was added to models that use sliding window, but this bug was not yet fixed.

### CODE REVIEW

**Comment by ehuaa** [comment]
Hi, @fxmarty ,thanks for your great work! I think you can add some tests i mentioned in https://github.com/huggingface/transformers/pull/29407, to check if the result with sliding window in SDPA is the same as flashattention2.

**Comment by fxmarty** [comment]
@ehuaa Thank you, for sure will do!

Runing mistral, mixtral, starcoder2 tests, those fail but are already failing on main:
```
FAILED tests/models/mistral/test_modeling_mistral.py::MistralIntegrationTest::test_speculative_generation - AssertionError: 'My f[19 chars]is 100% Sriracha. I love the heat, the tang and the fact costs' != 'My f[19 chars]is 100% Sriracha. I love the heat, the ...
FAILED tests/models/mixtral/test_modeling_mixtral.py::MixtralIntegrationTest::test_small_model_logits - AssertionError: The values for attribute 'dtype' do not match: torch.float16 != torch.float32.
FAILED tests/models/mixtral/test_modeling_mixtral.py::MixtralIntegrationTest::test_small_model_logits_batched - AssertionError: The values for attribute 'device' do not match: cuda:0 != cpu.
FAILED tests/models/starcoder2/test_modeling_starcoder2.py::Starcoder2IntegrationTest::test_starcoder2_batched_generation_4bit - AssertionError: Lists differ: ['Hel[110 chars]t is related to the topic of "How to make a ga[179 chars]ute'] != ['Hel[110 chars]t is aimed at creating a...
FAILED tests/models/starcoder2/test_modeling_starcoder2.py::Starcoder2IntegrationTest::test_starcoder2_batched_generation_eager - AssertionError: Lists differ: ['Hel[181 chars]I am currently working on', "def hello_world()[114 chars]app"] != ['Hel[181 chars]I am looking for a', "de...
```

File: src/transformers/modeling_attn_mask_utils.py [hunk 1]
```diff
@@ -316,7 +316,12 @@ def _prepare_4d_causal_attention_mask_for_sdpa(
         or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
     )
 
-    if attention_mask is not None:
+    ignore_causal_mask = False
+
+    if attention_mask is None:
+        if sliding_window is None or key_value_length < sliding_window:
+            ignore_causal_mask = not is_tracing
+    elif sliding_window is None or key_value_length < sliding_window:
```
Comment (ArthurZucker) [inline src/transformers/modeling_attn_mask_utils.py:324]: There are basically 2 cases:
1. You ignore the causal mask 
2. You don't ignore it. 
Code is really not super clear but we will refactor this soon anyways.

**Review by ArthurZucker — APPROVED** [review]
Don't know if I mentioned it offline, we'll refactor this to a single function without inheritance similar to `update_causal_mask`. See Recurrent Gemma, as it supports sliding window! 

Thanks for re-enabling sliding window.

**Pull request merged**